namer/marathon/src/main/scala/io/buoyant/namer/marathon/Authenticator.scala

```diff
@@ -0,0 +1,109 @@
+package io.buoyant.namer.marathon
+
+import com.twitter.finagle.{Failure, http}
+import com.twitter.util.{Future, Promise, Return, Throw, Time}
+import io.buoyant.marathon.v2.Api
+import java.net.URL
+import java.util.concurrent.atomic.AtomicReference
+import pdi.jwt.{Jwt, JwtAlgorithm}
+import scala.annotation.tailrec
+
+object Authenticator {
+
+  case class AuthRequest(
+    loginEndpoint: String,
+    uid: String,
+    privateKey: String,
+    algorithm: JwtAlgorithm = JwtAlgorithm.RS256
+  ) {
+    val path: String = new URL(loginEndpoint).getPath
+    val jwt: String = {
+      val token = Jwt.encode(s"""{"uid":"$uid"}""", privateKey, algorithm)
+      s"""{"uid":"$uid","token":"$token"}"""
+    }
+  }
+  private[marathon] case class UnauthorizedResponse(rsp: http.Response) extends Throwable
+  private[this] case class AuthToken(token: Option[String])
+  private[this] val closedException = Failure("closed").flagged(Failure.Interrupted)
+  private[this] val closedExceptionF = Future.exception(closedException)
+  private[this] val missingTokenException = Failure("missing token")
+  private[this] val missingTokenExceptionF = Future.exception(missingTokenException)
+
+  class Authenticated(client: Api.Client, authRequest: AuthRequest) extends Api.Client {
+    private[this] sealed trait State
+    private[this] object Init extends State
+    private[this] case class Authenticating(token: Future[String]) extends State
+    private[this] object Closed extends State
+    private[this] val state = new AtomicReference[State](Init)
+
+    final def apply(req: http.Request): Future[http.Response] = state.get match {
+      case Init =>
+        val tokenP = new Promise[String]
+        if (state.compareAndSet(Init, Authenticating(tokenP))) {
+          tokenP.become(getToken())
+          tokenP.flatMap(issueAuthed(req, _))
+        } else apply(req) // try again on race
+
+      case s0@Authenticating(tokenF) =>
+        tokenF.flatMap(issueAuthed(req, _)).rescue {
+          case UnauthorizedResponse(rsp) =>
+            // If a request is unauthorized even though we have a
+            // token, it may have expired. If another request hasn't
+            // already started authenticating, get a new token and
+            // reissue the request at most once.
```

Comment:
@jgehrcke that's definitely a nice optimization, thanks for pointing that out! will either change it here or in a subsequent PR.